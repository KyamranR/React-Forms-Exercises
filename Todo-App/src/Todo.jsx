import { useState } from "react";

function Todo({ id, task, removeTodo, updateTodo }) {
  const [isEditing, setIsEditing] = useState(false);
  const [editTask, setEditTask] = useState(task);

  const toggleEdit = () => setIsEditing((edit) => !edit);

  const handleChange = (e) => {
    setEditTask(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    updateTodo(id, editTask);
    setIsEditing(false);
  };
  return (
    <div>
      {isEditing ? (
        <form onSubmit={handleSubmit}>
          <input
            type="text"
            value={editTask}
            onChange={handleChange}
            placeholder="Edit task"
          />
          <button type="submit">Save</button>
        </form>
      ) : (
        <div>
          <span>{task}</span>
          <button onClick={toggleEdit}>Edit</button>
          <button onClick={() => removeTodo(id)}>X</button>
        </div>
      )}
    </div>
  );
}

export default Todo;
