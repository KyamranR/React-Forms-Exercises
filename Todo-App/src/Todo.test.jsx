import { fireEvent, render } from "@testing-library/react";
import Todo from "./Todo";
import { vi } from "vitest";

test("renders without crashing", () => {
  render(<Todo id="1" task="Test Todo" removeTodo={() => {}} />);
});

test("matches snapshot", () => {
  const { asFragment } = render(
    <Todo id="1" task="Test Todo" removeTodo={() => {}} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("calls removeTodo on button click", () => {
  const removeTodo = vi.fn();
  const { getByText } = render(
    <Todo id="1" task="Test todo" removeTodo={removeTodo} />
  );

  fireEvent.click(getByText("X"));
  expect(removeTodo).toHaveBeenCalledWith("1");
});

test("displays edit form and updates task", () => {
  const updateTodo = vi.fn();
  const { getByText, getByDisplayValue, queryByDisplayValue } = render(
    <Todo
      id="1"
      task="Test Todo"
      removeTodo={() => {}}
      updateTodo={updateTodo}
    />
  );

  fireEvent.click(getByText("Edit"));
  expect(getByDisplayValue("Test Todo")).toBeInTheDocument();

  const input = getByDisplayValue("Test Todo");
  fireEvent.change(input, { target: { value: "Updated Todo" } });
  fireEvent.click(getByText("Save"));

  expect(updateTodo).toHaveBeenCalledWith("1", "Updated Todo");

  expect(queryByDisplayValue("Updated Todo")).not.toBeInTheDocument();
});
