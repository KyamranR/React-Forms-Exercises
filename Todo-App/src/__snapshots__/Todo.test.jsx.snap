// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <span>
        Test Todo
      </span>
      <button>
        Edit
      </button>
      <button>
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
